src/middlewared/middlewared/plugins/pool.py

```diff
@@ -4260,33 +4288,69 @@ def __handle_zfs_set_property_error(self, e, properties_definitions):
             )
 
         quota_list = []
+        dataset_quotas = {}
+
         for i, q in enumerate(data):
             quota_type = q["quota_type"].lower()
-            if not q["id"].isdigit():
-                try:
-                    await self.middleware.call(f'{quota_type}.get_{quota_type}_obj',
-                                               {f'{quota_type}name': q["id"]})
-                except Exception:
+            if q["quota_type"] == 'DATASET':
+                if q['id'] not in ['QUOTA', 'REFQUOTA']:
                     verrors.add(
                         f'quotas.{i}.id',
-                        f'{quota_type} {q["id"]} is not valid.'
+                        f'id for quota_type DATASET must be either "QUOTA" or "REFQUOTA"'
                     )
+                    continue
+
+                if dataset_quotas.get(q['id'].lower(), None) is not None:
+                    verrors.add(
+                        f'quotas.{i}.id',
+                        f'Setting multiple values for {q["id"]} for quota_type "DATASET" is not permitted'
+                    )
+                    continue
+
+                dataset_quotas.update({
+                    q['id'].lower(): q['quota_value']
+                })
+
+                continue
+
+            if q["quota_type"] not in ['PROJECT', 'PROJECTOBJ']:
```

Comment:
Can `quota_type` ever be `PROJECT/PROJECTOBJ` ? I think not ?

Reply:
At the time I was writing this, I was considering future expansion of available quota types to include these as well. Project quotas will probably require a separate design doc and ticket, but I thought it's not a bad idea to at least keep some of the code here for future use. That said, I can remove if it's a problem.